content/advent-2016/exposing-go-on-the-internet.md

```diff
@@ -0,0 +1,208 @@
+
+[heartbleed]: https://github.com/FiloSottile/Heartbleed/commit/4a3332ca1dc07aedf24b8540857792f72624cdf7
+[tcpKeepAliveListener]: https://github.com/golang/go/blob/61db2e4efa2a8f558fd3557958d1c86dbbe7d3cc/src/net/http/server.go#L3023-L3039
+
+### ServeMux
+
+Package level functions like `http.Handle[Func]` (and maybe your web framework) register handlers on the global `http.DefaultServeMux` which is used if `Server.Handler` is nil. You should avoid that.
+
+Any package you import, directly or through other dependencies, has access to `http.DefaultServeMux` and might register routes you don't expect. 
+
+For example if any package somewhere in the tree imports `net/http/pprof` clients will be able to get CPU profiles for your application. You can still use `net/http/pprof` by registering [its handlers][pprof] manually.
+
+Instead, instantiate an `http.ServeMux` yourself, register handlers on it, and set it as `Server.Handler`. Or set whatever your web framework exposes as `Server.Handler`.
+
+[pprof]: https://github.com/golang/go/blob/1106512db54fc2736c7a9a67dd553fc9e1fca742/src/net/http/pprof/pprof.go#L67-L71
+
+### Logging
+
+`net/http` does a number of things before yielding control to your handlers: [`Accept`s the connections][accept], [runs the TLS Handshake][handshake], ...
+
+If any of these go wrong a line is written directly to `Server.ErrorLog`. Some of these, like timeouts and connection resets, are expected on the open Internet. It's not clean, but you can intercept most of those and turn them into metrics by matching them with regexes from the Logger Writer, thanks to this guarantee:
+
+> Each logging operation makes a single call to the Writer's Write method.
+
+To abort from inside a Handler without logging a stack trace you can either `panic(nil)` or in Go 1.8 `panic(http.ErrAbortHandler)`.
+
+[accept]: https://github.com/golang/go/blob/1106512db54fc2736c7a9a67dd553fc9e1fca742/src/net/http/server.go#L2631-L2653
+[handshake]: https://github.com/golang/go/blob/1106512db54fc2736c7a9a67dd553fc9e1fca742/src/net/http/server.go#L1718-L1728
+
+### Metrics
+
+A metric you'll want to monitor is the number of open file descriptors. [Prometheus does that by using the `proc` filesystem][proc].
+
+If you need to investigate a leak, you can use the `Server.ConnState` hook to get more detailed metrics of what stage the connections are in. However, note that there is no way to keep a correct count of Active connections without state, so you'll need to maintain a `map[net.Conn]ConnState`.
+
+[proc]: https://github.com/prometheus/client_golang/blob/master/prometheus/process_collector.go
```

Comment:
Did you really want to link to the source code here, I presume there wasn't some clearer docs to link to instead?